stdlib/list.ml

```diff
@@ -253,6 +262,16 @@ let rec combine l1 l2 =
   | (a1::l1, a2::l2) -> (a1, a2) :: combine l1 l2
   | (_, _) -> invalid_arg "List.combine"
 
+let rec product l1 l2 =
+  match l1 with
+  | [] -> []
+  | a1::l1 -> product_aux a1 l2 l1 l2
+
+and product_aux a1 l2 next1 original2 =
+  match l2 with
+  | [] -> product next1 original2
+  | a2::l2 -> (a1, a2) :: product_aux a1 l2 next1 original2
```

Comment:
`product` needs to either be tail-recursive or flagged as "not tail recursive (product of the lengths of the arguments)" in the documentation.

Reply:
Which of those do you think is the better option? I'm basically happy with either approach.

Reply:
I'm not sure how often people will want to use `product` with large lists,  but the limit does seem quite low — `product x x` will fail with a list of a few hundred elements  — so I'd be inclined to make the function tail-recursive.

Reply:
+1 for tail-recursive.

Reply:
> product x x will fail with a list of a few hundred elements

If we make it tail recursive, it is `List.map` that will fail on the result.  I think to remember that the "official" position was that the problem is not with non-tail recursive function in `List`, but with long lists per se.  I don't agree, and I believe that as long as the system doesn't know how to deal nicely with non-tail recursion (e.g. by extending the stack on demand), the Stdlib implementation should make efforts to protect the programmer from stack overflows, but hopefully without sacrificing performance too much.